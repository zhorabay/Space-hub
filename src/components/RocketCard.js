import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { handleRocket } from '../redux/rockets/rockets';
import '../styles/Rockets.css';

function RocketCard({ props }) {
  const {
    id, img, name, description, reserved,
  } = props;
  const dispatch = useDispatch();

  const getButton = (reserved, btn) => {
    let button;
    if (btn === 'reserve') {
      button = reserved ? (
        <button className="cancel-reservation" type="button" onClick={() => dispatch(handleRocket(id))}>Cancel Reservation</button>
      ) : (
        <button className="reserve-rocket" type="button" onClick={() => dispatch(handleRocket(id))}>Reserve Rocket</button>
      );
    }
    if (btn === 'reservation') {
      button = reserved ? (
        <button className="reserved" type="button">Reserved</button>
      ) : (
        <button className="not-reserved" type="button">Not reserved</button>
      );
    }
    return button;
  };

  return (
    <div className="rocket-card">
      <img className="rocket-img" src={img} alt="rocket" />
      <div className="rocket-body">
        <h2 className="rocket-name">{name}</h2>
        <p className="rocket-description">
          {getButton(reserved, 'reservation')}
          {description}
        </p>
        {getButton(reserved, 'reserve')}
      </div>
    </div>
  );
}

RocketCard.propTypes = {
  props: PropTypes.shape({
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
  }).isRequired,
  id: PropTypes.string.isRequired,
  img: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  reserved: PropTypes.bool.isRequired,
};

export default RocketCard;
