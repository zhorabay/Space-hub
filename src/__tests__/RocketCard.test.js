import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import '@testing-library/jest-dom/extend-expect';
import RocketCard from '../components/RocketCard';

const mockStore = configureStore([]);

describe('RocketCard Component', () => {
  it('renders the component with reserved badge if reserved is true', () => {
    const rocket = {
      id: '1',
      name: 'Rocket 1',
      image: 'rocket1.jpg',
      description: 'Description for Rocket 1',
      reserved: true,
    };

    const store = mockStore();

    render(
      <Provider store={store}>
        <RocketCard props={rocket} />
      </Provider>,
    );

    const reservedBadge = screen.getByText(/Reserved/i);
    expect(reservedBadge).toBeInTheDocument();
    expect(reservedBadge).toMatchSnapshot();
  });

  it('renders the component without reserved badge if reserved is false', () => {
    const rocket = {
      id: '2',
      name: 'Rocket 2',
      image: 'rocket2.jpg',
      description: 'Description for Rocket 2',
      reserved: false,
    };

    const store = mockStore();

    render(
      <Provider store={store}>
        <RocketCard props={rocket} />
      </Provider>,
    );

    const reservedBadge = screen.queryByText(/Reserved/i);
    expect(reservedBadge).toBeNull();
  });

  it('calls reserveRocket action when Reserve button is clicked', () => {
    const rocket = {
      id: '3',
      name: 'Rocket 3',
      image: 'rocket3.jpg',
      description: 'Description for Rocket 3',
      reserved: false,
    };

    const store = mockStore();

    render(
      <Provider store={store}>
        <RocketCard props={rocket} />
      </Provider>,
    );

    const reserveButton = screen.getByText(/Reserve rocket/i);
    fireEvent.click(reserveButton);
  });
});
