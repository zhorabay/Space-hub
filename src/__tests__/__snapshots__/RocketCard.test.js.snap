// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RocketCard Component renders the component with reserved badge if reserved is true 1`] = `
<span
  class="reserved-badge"
>
  Reserved
</span>
`;
